Use sendWithAsyncReply for starting and requesting additional items for a drag session

achristensen07 opened the pull request:
#### 243e10f98b11eb286da69baf032739f37014232c
<pre>
Use sendWithAsyncReply for starting and requesting additional items for a drag session
<a href="https://bugs.webkit.org/show_bug.cgi?id=292565">https://bugs.webkit.org/show_bug.cgi?id=292565</a>
<a href="https://rdar.apple.com/150717298">rdar://150717298</a>

Reviewed by Wenson Hsieh.

Use a CompletionHandler instead of DragStartRequestResult::Delayed.

* Source/WebCore/page/DragActions.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didHandleDragStartRequest): Deleted.
(WebKit::PageClientImpl::didHandleAdditionalDragItemsRequest): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _dragInteraction:itemsForAddingToSession:withTouchAtPoint:completion:]):
(-[WKContentView _dragInteraction:prepareForSession:completion:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDragStart):
(WebKit::WebPageProxy::requestAdditionalItemsForDragSession):
(WebKit::WebPageProxy::didHandleDragStartRequest): Deleted.
(WebKit::WebPageProxy::didHandleAdditionalDragItemsRequest): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):

Canonical link: <a href="https://commits.webkit.org/294585@main">https://commits.webkit.org/294585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1775b064803d61aa1478f1a16de3ca049a0110a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107550 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53025 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22366 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30564 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77899 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34890 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17320 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92423 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58236 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/101863 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10451 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52383 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10522 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109925 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29521 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86879 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88619 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31293 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9012 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16626 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29449 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/34751 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29260 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32583 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30819 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -530,12 +530,12 @@ void HTMLModelElement::renderingAbruptlyStopped()
 }
 
-bool HTMLModelElement::tryAnimateModelToFitPortal()
+void HTMLModelElement::tryAnimateModelToFitPortal(CompletionHandler<void(bool)>&& completionHandler, bool handledDrag)
```
Optional nit — I think we generally make completion handlers the second argument?
```suggestion
void HTMLModelElement::tryAnimateModelToFitPortal(bool handledDrag, CompletionHandler<void(bool)>&& completionHandler)
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -813,11 +813,11 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
-    if (subframe && subframe->eventHandler().tryToBeginDragAtPoint(adjustedClientPosition, adjustedGlobalPosition) == DragStartRequestResult::Started)
-        return DragStartRequestResult::Started;
+    if (subframe)
+        return subframe->eventHandler().tryToBeginDragAtPoint(adjustedClientPosition, adjustedGlobalPosition, WTFMove(completionHandler));
```
I _think_ this part is probably fine, given that starting a drag over the non-draggable part of the inner iframe in https://whsieh.github.io/examples/draggable-iframe doesn't start a drag in the main frame before this change. Would be good to double check that dragging within the inner frame on that test page still works after your change.

whsieh commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -1923,11 +1923,19 @@ - (void)updateLayoutIgnorePendingStyleSheets
+    struct BoolResult : public RefCounted<BoolResult> {
+        std::optional<bool> result;
+    };
+    Ref result = adoptRef(*new BoolResult());
```
Perhaps you can just use `WTF::Box<std::optional<bool>>` instead?

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -530,12 +530,12 @@ void HTMLModelElement::renderingAbruptlyStopped()
+        return m_modelPlayer->animateModelToFitPortal(WTFMove(completionHandler));
 
-    return false;
+    return completionHandler(handledDrag);
```
Nit: is the `return` needed?

achan00 commented on Source/WebCore/Modules/model-element/ModelPlayer.cpp:
```diff
@@ -163,9 +163,9 @@ void ModelPlayer::renderingAbruptlyStopped()
+void ModelPlayer::animateModelToFitPortal(CompletionHandler<void(bool)>&& completionHandler)
 {
-    return false;
+    return completionHandler(false);
```
Ditto.

achristensen07 committed:
Use sendWithAsyncReply for starting and requesting additional items for a drag session
https://bugs.webkit.org/show_bug.cgi?id=292565
rdar://150717298

Reviewed by Wenson Hsieh.

Use a CompletionHandler instead of DragStartRequestResult::Delayed.

* Source/WebCore/page/DragActions.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didHandleDragStartRequest): Deleted.
(WebKit::PageClientImpl::didHandleAdditionalDragItemsRequest): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _dragInteraction:itemsForAddingToSession:withTouchAtPoint:completion:]):
(-[WKContentView _dragInteraction:prepareForSession:completion:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDragStart):
(WebKit::WebPageProxy::requestAdditionalItemsForDragSession):
(WebKit::WebPageProxy::didHandleDragStartRequest): Deleted.
(WebKit::WebPageProxy::didHandleAdditionalDragItemsRequest): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):

Canonical link: https://commits.webkit.org/294585@main
